src/machi_merkle_tree_mgr.erl

```diff
@@ -0,0 +1,335 @@
+                    {update, Filename, Offset, Length, Csum}, ?TIMEOUT).
+
+-spec fetch ( FluName :: atom(),
+             Filename :: string() ) -> {ok, [ Data :: binary() ]}.
+%% @doc Returns the entire merkle tree for the given filename.
+%% {@link fetch/3}
+fetch(FluName, Filename) ->
+    fetch(FluName, Filename, all).
+
+-spec fetch(  FluName :: atom(),
+             Filename :: string(),
+                Level :: 'root' | 'all' 
+           ) -> {ok, [ Data :: binary() ]}.
+%% @doc Return the current merkle tree for the given filename. 
+%% If `root' is specified, returns a list with 1 element, the root
+%% checksum of the tree.  If `all' is specified, returns a list
+%% with all levels.
+fetch(FluName, Filename, Level) ->
+    gen_server:call(make_merkle_tree_mgr_name(FluName),
+                    {fetch, Filename, Level}, ?TIMEOUT).
+
+%% gen_server callbacks
+init({FluName, DataDir, Options}) ->
+    Tid = ets:new(make_merkle_tree_mgr_name(FluName), [{keypos, 2}, {read_concurrency, true}]),
+    case proplists:get_value(no_load, Options, false) of
+        true ->
+            ok;
+        false ->
+            handle_load(Tid, DataDir)
+    end,
+    {ok, #state{fluname=FluName, datadir=DataDir, tid = Tid}}.
+
+handle_call({fetch, Filename, Level}, _From, S = #state{ tid = Tid }) ->
+    Res = handle_fetch(Tid, Filename, Level),
+    {reply, {ok, [ Res ]}, S};
+handle_call(Req, _From, State) ->
+    lager:warning("Unknown call: ~p", [Req]),
+    {reply, whoaaaaaaaaaaaa, State}.
+
+handle_cast({initialize, Filename}, S = #state{ datadir = D, tid = Tid }) ->
+    load_filename(Tid, D, Filename),
+    {noreply, S};
+
+handle_cast({update, Filename, Offset, Length, Csum}, S = #state{ tid = Tid }) ->
+    %% XXX FIXME: Not sure about the correctness of this
+    insert(Tid, Filename, {Offset, Length, Csum}),
+    {noreply, S};
+
+handle_cast(Cast, State) ->
+    lager:warning("Unknown cast: ~p", [Cast]),
+    {noreply, State}.
+
+handle_info(Req, State) ->
+    lager:warning("Unknown info message: ~p", [Req]),
+    {noreply, State}.
+
+terminate(Reason, #state{fluname = F}) ->
+    lager:debug("Shutting down merkle tree manager for FLU ~p because ~p", 
+                [F, Reason]),
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%% private
+
+make_merkle_tree_mgr_name(FluName) ->
+    list_to_atom(atom_to_list(FluName) ++ "_merkle_tree_mgr").
+
+handle_load(Tid, DataDir) ->
+    Files = get_files(DataDir),
+    lists:foreach(fun(F) -> load_filename(Tid, DataDir, F) end, Files).
+
+get_files(_DataDir) ->
+    [].
+
+load_filename(Tid, DataDir, Filename) ->
+    CsumFile = machi_util:make_checksum_filename(DataDir, Filename),
+    case file:read_file(CsumFile) of
+        {error, enoent} -> 
+            insert(Tid, Filename, {0, ?MINIMUM_OFFSET, unwritten});
+        {ok, Bin} ->
+            load_bin(Tid, Filename, Bin);
+        Error ->
+            throw(Error)
+    end.
+
+load_bin(Tid, Filename, Bin) ->
```

Comment:
Folding over checksum table should be abstractized like `machi_csum_table:fold_entries(T, Fun, InitAcc)` . This is in case for future migration to other key-value stores and to keep `machi_csum_table`-related features testable.